### PR TITLE
checker(dm): fix deadlock when check-task/start-task

### LINKED ISSUE #5759: start-task/check-task blocked and printed few logs
### What did you do?

https://asktug.com/t/topic/692986

- check-task automatically before running

### What did you expect to see?

check-task successfully and start to run.

### What did you see instead?

dmmaster was blocked for long without response

### Versions of the cluster

6.0.0

### current status of DM cluster (execute `query-status <task-name>` in dmctl)

```console
(paste current status of DM cluster here)
```

### PR DESCRIPTION
**Pull request by lance6716**
### What problem does this PR solve?

Issue Number: close #5759 

### What is changed and how it works?

ShardingTablesChecker: return an error so errgroup will cancel the context
TablesChecker: start consumer of channel earlier

#### Tests 

 - Unit test
 - Integration test

### Release note 

```release-note
Fix a bug that start-task/check-task may stuck.
```

### CODE REVIEW

File: dm/tests/check_task/conf/task-sharding.yaml [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+target-database:       # 下游数据库实例配置
+  host: "127.0.0.1"
+  port: 4000
+  user: "root"
+  password: ""         # 如果密码不为空，则推荐使用经过 dmctl 加密的密文
+
+## ******** 功能配置集 **********
+block-allow-list:        # 上游数据库实例匹配的表的 block-allow-list 过滤规则集，如果 DM 版本 <= v2.0.0-beta.2 则使用 black-white-list
+  bw-rule-1:             # 黑白名单配置的名称
+    do-dbs: ["tug"] # 迁移哪些库
```
Comment (lance6716) [inline dm/tests/check_task/conf/task-sharding.yaml:14]: no tables under this db? @buchuitoudegou

File: dm/tests/check_task/run.sh [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+    run_dm_ctl $WORK_DIR "127.0.0.1:$MASTER_PORT" \
+        "check-task $cur/conf/task-noshard.yaml" \
+        "\"state\": \"fail\"" 1
+}
+
+function test_check_task_fail_no_block_forsharding() {
+    prepare_many_tables
+    run_dm_ctl $WORK_DIR "127.0.0.1:$MASTER_PORT" \
+        "check-task $cur/conf/task-noshard.yaml" \
+        "pre-check is passed" 1
```
Comment (lance6716) [inline dm/tests/check_task/run.sh:43]: this also should fail

File: dm/pkg/checker/table_structure.go [hunk 3]
```diff
@@ -120,10 +121,12 @@ func (c *TablesChecker) Check(ctx context.Context) *Result {
 			return c.checkTable(checkCtx)
 		})
 	}
-
-	dispatchTableItem(checkCtx, c.tableMap, c.inCh)
+	// start consuming results before dispatching
+	// or the dispatching thread could be blocked when
```
Comment (D3Hunter) [inline dm/pkg/checker/table_structure.go:125]:
```suggestion
	// else the dispatching thread could be blocked when
```

File: dm/pkg/checker/utils.go [hunk 4]
```diff
@@ -125,7 +125,13 @@ func IsTiDBFromVersion(version string) bool {
 	return strings.Contains(strings.ToUpper(version), "TIDB")
 }
 
+var processedError = errors.New("error has been processed, can skip it")
+
 func markCheckError(result *Result, err error) {
+	if err == nil || err == processedError {
+		return
+	}
+
```
Comment (D3Hunter) [inline dm/pkg/checker/utils.go:134]: should revert too

File: dm/tests/check_task/run.sh [hunk 5]
```diff
@@ -0,0 +1,56 @@
+#!/bin/bash
```
Comment (D3Hunter) [inline dm/tests/check_task/run.sh:1]: maybe move it into dmctl_basic/*/check_task.sh

**Review by D3Hunter — COMMENTED** [review]
rest lgtm

File: dm/pkg/checker/table_structure.go (hunk 3 above)
Comment (buchuitoudegou) [reply dm/pkg/checker/table_structure.go:125]: or => otherwise

File: dm/tests/check_task/run.sh (hunk 5 above)
Comment (buchuitoudegou) [reply dm/tests/check_task/run.sh:1]: Task checking isn't solely relevant to `dmctl` and it has incurred a noticeable number of issues (including connection checking etc.). So I'm planning to decouple these test cases from others and test them as a whole.

**Review by D3Hunter — APPROVED** [review]

File: dm/pkg/checker/table_structure.go [hunk 6]
```diff
@@ -456,7 +461,9 @@ func (c *ShardingTablesChecker) checkShardingTable(ctx context.Context, r *Resul
 				r.Extra = fmt.Sprintf("error on sharding %s", c.targetTableID)
 				r.Instruction = "please set same table structure for sharding tables"
```
Comment (lance6716) [inline dm/pkg/checker/table_structure.go:462]: these fields (except Errors) may be overwritten in next loop, at that time output is misleading. Maybe we return on first error is OK.

@GMHDBJD

File: dm/tests/check_task/conf/source2.yaml [hunk 7]
```diff
@@ -0,0 +1,12 @@
+
```
Comment (lance6716) [inline dm/tests/check_task/conf/source2.yaml:1]: we only use one source

**Review by GMHDBJD — APPROVED** [review]
LGTM

**Pull request merged**